sessions/2018/october/Session_14.md

```diff
@@ -0,0 +1,118 @@
+```
+@SpringBootApplication
+public class DemoApplication {
+
+	public static void main(String[] args) {
+		SpringApplication.run(DemoApplication.class, args);
+	}
+}
+```
+
+- ##### [@RestController][4], [@RequestMapping][5]
+
+Controllers is where the logic of the web application happends and where maps (HTTP request) are processed. Spring looks for `@RestController` (and other `@Controller`) annotations when it starts up.
+
+`@RequestMapping` indicates Spring to map HTTP request to the controller methods:
+
+```java
+@RestController
+@RequestMapping("/api")
+public class DemoController {
+  ...
+}
+```
+
+- ##### [@Service][6]
+
+Classes denoted with `@Service` annotation are intented to perform bussiness logic:
+```java
+@Service
+public class DemoService {
+  ...
+}
+```
+Tasks that controller delegates such as validation are common in this kind of classes, whereas database callings or handling HTTP endpoints aren't.
+
+##### Note about databases:
+In Spring, the data we get back from the database is the model.
+
+There is _classes for getting the data from the database_, these classes (making the database calls) are annotated with [`@Repository`][6] and might have complicated methods for getting the right data out. E.g: `getProjectsForUsers(List<Long> userIds, int pageCount, int pageIndex)`.
+
+Also exists _data classes_ that has fields, getters and setters. No complicated methods
+
+In other frameworks like Ruby on Rails, the _data classes_ and the _classes for getting the data from the database_ are the same classes, so they could be called indistinguishably "the model" in the MVC approach.
+
+- ##### [@Autowired][7]
+
+`@Autowired` annotations makes Spring inject the field that thinks is right based on the type. Only classes registered as `@Service` or `@RestController` or other [spring annotations][2]  can be injected.
+```java
+@RestController
+@RequestMapping("/api")
+public class DemoController {
+    @Autowired
+    private DemoService service;
+    ...
+}
+```
+
+- ##### [@GetMapping][8], [@PostMapping][8]  
+
+`@GetMapping` annotation is for handle `GET` HTTP requests and `@PostMapping` for receive data through `POST` HTTP request. In order to test `POST` endpoints, an application such as [Postman][10] must be used.
+
+
+###  Homework
+
+Based in the provided Roger example. Make an application that handle next maps(routes):
+
+| Map | Method | Description |
+|---|---|---|
+|`/api/customer` |GET| Give me all customers. Hardcoded|
+|`/api/book`| POST | Save to an in-memory map in a service. Return a json with the ID of the newly created book.|
+|`/api/:id` |GET| Get by ID, should return a previously posted book|
+
+Also, write Integration tests
+
+**TIP**: Following, an example for the POST /book map:
+```java
+@GetMapping("/{id}")
+public MyPojo getData(@PathVariable("id") String id) {
```

Comment:
Could you write in your own words what `@PathVariable` does?

Reply:
Done, thanks!